### PR TITLE
Backing image enhancement

### PR DESCRIPTION
**Pull request by shuo-wu**
longhorn/longhorn#2530

### CODE REVIEW

**Comment by shuo-wu** [comment]
Adding error message reporting and retry backoff. Convert to draft for now.

File: datastore/longhorn.go
Comment (joshimoo) [inline datastore/longhorn.go:1437]: NIT: I like to add a generic list function that gets passed a selector, see the refactor I done for the `listBackingImageManagers(selector labels.Selector)` as an example. This way both the list Everything an ByNode both go through the same retrieval + copy code flow.

File: controller/backing_image_data_source_controller.go
Comment (joshimoo) [inline controller/backing_image_data_source_controller.go:612]: NIT: `return pod.Labels[types.GetLonghornLabelComponentKey()] == types.LonghornLabelBackingImageDataSource`

File: controller/backing_image_data_source_controller.go
Comment (joshimoo) [inline controller/backing_image_data_source_controller.go:826]: NIT: any reason not to use m.log for these?

File: controller/backing_image_data_source_controller.go
Comment (joshimoo) [inline controller/backing_image_data_source_controller.go:841]: NIT: this uses the old monitoring pattern, where the monitor does the status update outside of the reconcile loop.
It's okay for now, just need to remember to refactor it later when we do https://github.com/longhorn/longhorn/issues/2441

**Review by joshimoo — COMMENTED** [review]
I reviewed all of the code completely except the 3 controllers. Let's do an in person review of the controllers during tomorrows PR review meeting @shuo-wu

**Comment by joshimoo** [comment]
These PR's that add new controllers + crds are always pretty large.
We should consider splitting them into  2 PR's.
PR1: With CRD + Boilerplate Code (datastore, etc) and boilerplate Controller (enqueues, constructor) no sync logic yet.
PR2: actual implementation of the sync logic + monitoring etc.

This way PR1 can go early into the code base with a quick check review, since it's more or less just boilerplate.
Then PR2 can be reviewed in depth or via in person code walk through, since this is where actual code will run and do stuff.

What do you guys think @innobead @shuo-wu ?

File: util/util.go
Comment (shuo-wu) [reply util/util.go:432]: No. Will remove the naming.

File: types/types.go
Comment (shuo-wu) [reply types/types.go:167]: Sounds good. Will update it.

File: util/util.go
Comment (shuo-wu) [reply util/util.go:64]: No. I didn't realize the interval was too short. I will update it to 500ms.

BTW, what do you mean by `at some jitter around the time.sleep`?

File: engineapi/types.go
Comment (shuo-wu) [reply engineapi/types.go:122]: I think it's safe to remove it.
1. This field is never been used since we remove the backing image feature in v0.5.0. The backing image info is tracked by the new field `BackingImageName` after re-introducing backing image feature in v1.1.0.
2. Even if somehow the backup meta info json string contains this removed field, this field will be ignored during marshaling.

File: controller/backing_image_data_source_controller.go
Comment (shuo-wu) [reply controller/backing_image_data_source_controller.go:826]: No. I was just used to print the log via `logrus`. I will replace all `logrus` with `m.log`.

File: controller/backing_image_data_source_controller.go
Comment (shuo-wu) [reply controller/backing_image_data_source_controller.go:841]: Sure. I will add a `TODO` comment with this link first. If time allows, I will do refactor later.

**Comment by shuo-wu** [comment]
> These PR's that add new controllers + crds are always pretty large.
> We should consider splitting them into 2 PR's.
> PR1: With CRD + Boilerplate Code (datastore, etc) and boilerplate Controller (enqueues, constructor) no sync logic yet.
> PR2: actual implementation of the sync logic + monitoring etc.

It's a nice suggestion. But for this PR, splitting it now will lead to tons of extra modifications and commits adjustments, which will consume lots of time. Besides, there are 2 other PRs highly rely on this PR. I think we can follow this pattern next time. 

BTW, the controller modifications/commits are in the latter part of this PR. We can focus on these commits then.

**Comment by joshimoo** [comment]
@shuo-wu Agree I wasn't suggesting to do this to this PR, makes sense not to touch this one.

File: engineapi/types.go
Comment (shuo-wu) [reply engineapi/types.go:122]: Will call it out in the comit comment.

File: api/backingimage.go
Comment (shuo-wu) [inline api/backingimage.go:119]: Just return nil if the CR can not be found after retry

File: controller/backing_image_controller.go
Comment (shuo-wu) [inline controller/backing_image_controller.go:641]: Move the HA check to the cleanup function rather than this timestamp setting function

File: controller/backing_image_controller.go
Comment (shuo-wu) [inline controller/backing_image_controller.go:540]: Abstruct this part as a function to avoid duplicate code

File: controller/backing_image_controller.go
Comment (shuo-wu) [inline controller/backing_image_controller.go:400]: Use a variable with **an appropriate naming** to indicate the code logic.

**Review by joshimoo — COMMENTED** [review]
Had a PR review session with @shuo-wu we identified 2 issues, that would need to be fixed before merging. But the rest should be fine.

- Issue01: backup image listing should function even if a backup data source resource is missing (i.e. one missing ds should not stop all backup images from listing)

- Issue02: the way cleanup of a no longer referenced image is handled needs to be changed, so there is no flip/flop on the last ref timestamp. Instead a free floating function should be designed that gets a set of input parameters that are necessary to produce a set of disks/images that are safe to remove. The reason for this is that this function can then be independently unit tested with different scenarios withhout having knowledge of the controller sync loops.

A similiar approach should in the future be used for replica scheduling decisions @c3y1huang

File: controller/node_controller.go
Comment (joshimoo) [inline controller/node_controller.go:955]: This is a good step forward with the addition of the unit tests.
In general I would have liked to see a split between evaluation/decision making (which to delete) & modification (operates on the prior set of decisions) then removes them from the resources.

This way Decision logic can be independently tested, but this is good enough :)
Thank you for adding it.

**Review by joshimoo — DISMISSED** [review]
LGTM Just gave it a quick look on the changes.

**Comment by joshimoo** [comment]
@innobead /hold merge @shuo-wu will merge this after he had a chance to do a full test.

**Comment by shuo-wu** [comment]
May need to update vendor longhorn/backing-image-manager before merge.

**Comment by joshimoo** [comment]
Had a review session with @shuo-wu LGTM, we identified some smaller changes.
And scheduled a PR review session for Monday, to follow up and get this merged.

**Review by innobead — APPROVED** [review]
In general, LGTM.

**Pull request merged**
